designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtypedata/ProcessingTypeDataProvider.scala

```diff
@@ -34,75 +32,126 @@ trait ProcessingTypeDataProvider[+T, +C] {
       )
     )
 
-  def all(implicit user: LoggedUser): Map[ProcessingType, T]
+  final def forType(processingType: ProcessingType)(implicit user: LoggedUser): Option[T] = allAuthorized
+    .get(processingType)
+    .map(_.getOrElse(throw new UnauthorizedError()))
+
+  def all(implicit user: LoggedUser): Map[ProcessingType, T] = allAuthorized.collect { case (k, Some(v)) =>
+    (k, v)
+  }
+
+  private def allAuthorized(implicit user: LoggedUser): Map[ProcessingType, Option[T]] = state.all.map {
+    case (k, ValueWithPermission(v, AnyUser)) => (k, Some(v))
+    case (k, ValueWithPermission(v, UserWithAccessRightsToCategory(category))) if user.can(category, Permission.Read) =>
+      (k, Some(v))
+    case (k, _) => (k, None)
+  }
 
   // TODO: We should return a generic type that can produce views for users with access rights to certain categories only.
   //       Thanks to that we will be sure that no sensitive data leak
-  def combined: C
+  def combined: C = state.getCombined()
 
-  def mapValues[Y](fun: T => Y): ProcessingTypeDataProvider[Y, C] = {
+  private[processingtypedata] def state: ProcessingTypeDataState[T, C]
 
-    new ProcessingTypeDataProvider[Y, C] {
+  def mapValues[TT](fun: T => TT): ProcessingTypeDataProvider[TT, C] =
+    new TransformingProcessingTypeDataProvider[T, C, TT, C](this, _.mapValues(fun))
 
-      override def forType(processingType: ProcessingType)(implicit user: LoggedUser): Option[Y] =
-        ProcessingTypeDataProvider.this.forType(processingType).map(fun)
+  def mapCombined[CC](fun: C => CC): ProcessingTypeDataProvider[T, CC] =
+    new TransformingProcessingTypeDataProvider[T, C, T, CC](this, _.mapCombined(fun))
 
-      override def all(implicit user: LoggedUser): Map[ProcessingType, Y] =
-        ProcessingTypeDataProvider.this.all.mapValuesNow(fun)
+}
 
-      override def combined: C = ProcessingTypeDataProvider.this.combined
-    }
+class TransformingProcessingTypeDataProvider[T, C, TT, CC](
+    observed: ProcessingTypeDataProvider[T, C],
+    transformState: ProcessingTypeDataState[T, C] => ProcessingTypeDataState[TT, CC]
+) extends ProcessingTypeDataProvider[TT, CC] {
 
+  private var stateValue: ProcessingTypeDataState[TT, CC] = transformState(observed.state)
+
+  override private[processingtypedata] def state: ProcessingTypeDataState[TT, CC] = synchronized {
+    val currentObservedState = observed.state
+    if (currentObservedState.stateIdentity != stateValue.stateIdentity) {
+      stateValue = transformState(currentObservedState)
+    }
+    stateValue
   }
 
-  def mapCombined[CC](fun: (=> C) => CC): ProcessingTypeDataProvider[T, CC] = {
+}
+
+object ProcessingTypeDataProvider {
 
-    new ProcessingTypeDataProvider[T, CC] {
+  val noCombinedDataFun: () => Nothing = () =>
```

Comment:
I would change this first occurrence of `()` to `Unit` in order to distinguish the `Unit` datatype from its only value which is `()`.

Reply:
But in this context, `()` is not a Unit, but empty parameters list :) It's just a function!